docs/processes/quality-assurance/testing-and-dev.md

```diff
@@ -0,0 +1,200 @@
+
+Testing tools include JEST (for javascript), Rails minitest (for Rails) (**etc, etc**) but look for something for your language.
+
+(Also mention mocking and expand)
+```
+
+### **Writing a failing test case**
+
+_Why are we writing a test to fail?_
+
+Having determined our test case from our acceptance criteria, we begin by writing some code that will fail our test. Why? Because we want to make sure that our test works; we don't want later checks to be useless because we wrote our test incorrectly and are having false positives returned.
+
+### **Writing code to pass your test**
+
+Once you've confirmed your test "works" by failing for code that doesn't pass your test, you can now write code to pass your test. This can also be an iterative process – first, starting with arbitrary values that you know will give the "expected" result (faking it to pass), before refining to be more broadly applicable to how your code needs to work in the live system.
+
+When you are satisfied that your code is complete, it is wise to challenge your test again with known failing code to reaffirm your testing is still functioning as expected.
+
+**What next?**
+
+### TBC refactoring intro \*\*\*
+
+_Below block to be cleaned up/expanded_
+
+```text
+Code quality:
+
+Is this reusable?
+
+Is this understandable?
+
+Is this readable?
+
+Can a third-party looking at your code easily understand what is going on?
+
+expand and intro into "refactoring code"
+```
+
+### **Refactoring Code**
+
+Refactoring involves changing the _internal_ structure to make it more easily understood, effective and/or cheaper to modify without changing the _external behaviour_ of the code.
+
+While writing a failing test case and then coding it to pass focuses on **requirements** (and adding **new functionality** ), refactoring focuses on **improving design**.
+
+**My code is complete and passing tests; is it time to refactor?**
+
+You don't need to assess your code for refactoring on every test, but if your find your code is:
+
+- repeating itself (or repeating functionality),
+- confusing to follow, or
+- you are getting stuck,
+
+these are good signs that it might be time to refactor.
+
+**How do I refactor?**
+
+Extract function refactoring is one common and intuitive approach to refactoring code: this involves comparing the code intention with the implementation. If the code is doing more than the intention, particularly if it is doing things that may be repeated elsewhere, it probably contains functions that should be split off.
+
+[Refactoring.com](http://refactoring.com/) provides useful reading that covers a number of refactoring techniques.
+
+**Further information about Test Driven Development**
+
+Deakin students have free access to LinkedIn Learning courses.
+
+A useful course on these general concepts is [Programming Foundations Test Driven Development](https://www.linkedin.com/learning/programming-foundations-test-driven-development-3/small-steps-to-great-things?u=2104084)
```

Comment:
**suggestion:** I think we can drop the `u` parameter as it is uniquely different for every user.

```suggestion
A useful course on these general concepts is [Programming Foundations Test Driven Development](https://www.linkedin.com/learning/programming-foundations-test-driven-development-3/small-steps-to-great-things).
```

Reply:
Thanks for picking that up, @tancnle !